router/core/header_rule_engine.go

```diff
@@ -37,158 +46,472 @@ var (
+					if value != "" {
+						request.Header.Set(rule.Rename, ctx.Request().Header.Get(name))
+						request.Header.Del(name)
+					} else if rule.Default != "" {
+						request.Header.Set(rule.Rename, rule.Default)
+						request.Header.Del(name)
 					}
+
+					continue
+				}
+
+				/**
+				 *	Propagate the header as is
+				 */
+				if slices.Contains(ignoredHeaders, name) {
+					continue
 				}
+				request.Header.Set(name, ctx.Request().Header.Get(name))
 			}
 		}
 	}
+}
 
-	return request, nil
+func (h *HeaderPropagation) applyResponseRuleMostRestrictiveCacheControl(res *http.Response, propagation *responseHeaderPropagation, rule *config.ResponseHeaderRule) {
+	ctx := res.Request.Context()
+	tracer := rtrace.TracerFromContext(ctx)
+	commonAttributes := []attribute.KeyValue{
+		otel.WgOperationProtocol.String(OperationProtocolHTTP.String()),
+	}
+
+	_, span := tracer.Start(ctx, "HeaderPropagation - RestrictiveCacheControl",
+		trace.WithSpanKind(trace.SpanKindInternal),
+		trace.WithAttributes(commonAttributes...),
+	)
+
+	cacheControlKey := "Cache-Control"
+	reqDir, _ := cachedirective.ParseRequestCacheControl(res.Request.Header.Get(cacheControlKey))
+	resDir, _ := cachedirective.ParseResponseCacheControl(res.Header.Get(cacheControlKey))
+	expiresHeader, _ := http.ParseTime(res.Header.Get("Expires"))
+	dateHeader, _ := http.ParseTime(res.Header.Get("Date"))
+	lastModifiedHeader, _ := http.ParseTime(res.Header.Get("Last-Modified"))
+
+	obj := &cachedirective.Object{
+		RespDirectives:         resDir,
+		RespHeaders:            res.Header,
+		RespStatusCode:         res.StatusCode,
+		RespExpiresHeader:      expiresHeader,
+		RespDateHeader:         dateHeader,
+		RespLastModifiedHeader: lastModifiedHeader,
+
+		ReqDirectives: reqDir,
+		ReqHeaders:    res.Request.Header,
+		ReqMethod:     res.Request.Method,
+
+		NowUTC: time.Now().UTC(),
+	}
+	rv := cachedirective.ObjectResults{}
+
+	cachedirective.CachableObject(obj, &rv)
+	cachedirective.ExpirationObject(obj, &rv)
+
+	span.SetAttributes(
+		otel.WgResponseCacheControlReasons.String(fmt.Sprint(rv.OutReasons)),
+		otel.WgResponseCacheControlWarnings.String(fmt.Sprint(rv.OutWarnings)),
+		otel.WgResponseCacheControlExpiration.String(rv.OutExpirationTime.String()),
+	)
+
+	propagation.m.Lock()
+	defer propagation.m.Unlock()
+
+	defaultResponseCache, _ := cachedirective.ParseResponseCacheControl(rule.Default)
+	defaultCacheControlObj := &cachedirective.Object{
+		RespDirectives: defaultResponseCache,
+	}
+
+	if propagation.previousCacheControl == nil {
+		if rule.Default != "" {
+			propagation.previousCacheControl = defaultCacheControlObj
+			propagation.header.Set(cacheControlKey, rule.Default)
+		} else {
+			propagation.previousCacheControl = obj
+			propagation.header.Set(cacheControlKey, res.Header.Get(cacheControlKey))
+			return
+		}
+	} else if rule.Default != "" && isMoreRestrictive(defaultCacheControlObj, propagation.previousCacheControl) {
```

Comment:
Maybe I miss something but the cache library has an api to determine the cacheability and expiration time of an object.

```go
cachedirective.CachableObject(obj, &rv)
cachedirective.ExpirationObject(obj, &rv)
```
Don't we only need to compare those values to calculate the more restrictive object?

Reply:
@StarpTech That's true, but the cache library doesn't have a direct api to compare two cachable objects, which is why I defined that function. More, when I was debugging it I haven't been able to find a case where the `ObjectResults` actually have relevant information (the Warnings and Reasons are always null in all of our test cases)